_docs/rfc/for_each_iteration.md

```diff
@@ -0,0 +1,218 @@
+`for_each` collection with `each.value` interpolated to the value for that item. Then, Terragrunt will run the command
+on each config in a loop.
+
+One challenge with this approach is that the state needs to be unique for each item in the `for_each` collection to
+avoid conflict in the state file. Since each call is a separate module call, having each one share state will result in
+each call stepping over each other. Therefore, it is important that this implementation **add a guard to ensure the
+remote state file is unique per iteration**.
+
+To support this, there should be a way to partially patch the `remote_state` config that is inherited from a parent. For
+example, we should be able to do the following:
+
+_Parent terragrunt.hcl_
+
+```
+remote_state {
+  backend = "s3"
+  config = {
+    bucket         = "my-terraform-state"
+    key            = "${path_relative_to_include()}/terraform.tfstate"
+    region         = "us-east-1"
+    encrypt        = true
+    dynamodb_table = "my-lock-table"
+  }
+}
+```
+
+_Child terragrunt.hcl_
+
+```
+for_each = ["us-east-1", "us-west-1"]
+
+remote_state {
+  # Added for backwards compatibility. This should default to false, so that existing config overrides completely as
+  # opposed to partially.
+  merge_parent = true
+
+  config = {
+    key = "dev/${each.value}/_global/aws-config/terraform.tfstate"
+  }
+}
+
+inputs = {
+  region = each.value
+}
+```
+
+In this setup, the `remote_state` block in the child `terragrunt.hcl` should be used to patch the parent `remote_state`
+block, as opposed to completely replacing it as it is now in the current behavior. This way, you can still keep your
+`remote_state` config DRY while only updating the relevant pieces to support iteration.
+
+##### Implementation detail
+
+To implement this, the following changes need to be made:
+
+- `ParseConfigFile`, `ParseConfigString`, `PartialParseConfigFile`, and `PartialParseConfigString` need to be updated to
+  return list of `TerragruntConfig` structs, as opposed to a single one.
+- Each caller of parsing functions need to accordingly handle the list of `TerragruntConfig` objects. As a part of this,
+  the caller should loop through each config to run the commands in the context of each iteration of the expanded loop.
+- The terragrunt workspace logic needs to be updated to ensure a separate module folder is created for each loop
+  iteration.
+- The parsing functions need to be updated to first parse the `for_each` attribute. Then, the parser should iterate each
+  item in the list and set the `each` variable accordingly as it parses the rest of the config. This means the parsing
+  order is now adjusted as:
+    - `for_each` and expand loop.
+    - `locals`
+    - `include`
+    - `dependency`
+    - Everything else
+- Support `merge_parent` in `remote_state`, which ultimately adjusts how `mergeConfigWithIncludedConfig` merges the
```

Comment:
Is there a way for us to check the user isn't accidentally overriding the config multiple times? Does `remote_state` config in Terragrunt become required to use this feature? Or do we read the `.terraform/terraform.tfstate` file? Seems a bit complicated.

Reply:
I think local state will work because I was imagining each iteration will have it's own directory in the cache, but yes this isn't immediately obvious.